### PR TITLE
feat: sqlalchemy integration

### PR DESCRIPTION
**Pull request by guotin**
Hey @matveypashkovskiy & @alexey-vyskubov 

Here's a version that uses SQLalchemy core. The entire test loop is bundled into a single transaction and primary keys of entries are kept in a dictionary during the run so they're not queried during every test function. Some column names are changed to be more descriptive. Also there were some cases where a source code file was inserted into the mapping database without any coverage and I removed that "feature". Otherwise the mapping database contains the same items. My optimizations helped a little but the performance is significantly worse than in master, taking 1.5x time in `pandas/tests/arithmetic` compared to current `master`. 

I had to do some refactoring as well to make `pylint` happy. There's still some weird bug that I can't figure out with `tox`. After adding a missing `__init__.py` to `pytest_rts/tests/`, the pytest run with tox won't pass but works without it. It's complaining that `pytest_rts.tests` doesn't exist. Maybe you have some idea why that is happening.

### CODE REVIEW

**Comment by matveypashkovskiy** [comment]
A question: how "1.5x time worse" figure was obtained? Was it a single run of all tests from `pandas/tests/arithmetic` or was it, say 10-20, runs and then mean was selected? Sometimes the second approach shows more realistic results. The same question applies to the performance of code from the master branch.

**Comment by matveypashkovskiy** [comment]
README.md needs some extra sentence about `--rts-db-conn` option

**Comment by guotin** [comment]
Force pushed to fix pytest error with `tox` and change mapping database column names as they are in `master`.

**Comment by guotin** [comment]
Now that I'm working on preparing a demo with pandas I remembered that also the "Update phase" for committed changes suffers from the slowdown that collecting coverage produces. It has slipped my mind during our discussions.

**Comment by matveypashkovskiy** [comment]
indeed, let's look at pytest-cov database as the next step

File: pytest_rts/pytest/mapper_plugin.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        """"Constructor calls database and Coverage.py initialization"""
+        self.cov = coverage.Coverage(data_file=None)
+        self.cov._warn_unimported_source = False
+        self.mappinghelper = mappinghelper
+        self.testfiles = {testfile[1] for testfile in self.mappinghelper.testfiles}
+        self.test_func_lines = None
+
+    def pytest_sessionstart(self, session):
+        """Start SQLalchemy transaction before test loop"""
+        del session
```
Comment (alexey-vyskubov) [inline pytest_rts/pytest/mapper_plugin.py:24]: I'm not sure I understand what happens here. Can you add a comment?

File: pytest_rts/utils/common.py [hunk 2]
```diff
@@ -162,3 +97,15 @@ def calculate_func_lines(src_code) -> Dict[str, Tuple[int, int]]:
...
+
+def filter_and_sort_pytest_items(test_set, pytest_items, runtimes) -> List[Item]:
+    """Selected pytest items based on found tests
+    ordered by their runtimes
+    """
+    selected = list(filter(lambda item: item.nodeid in test_set, pytest_items))
+    updated_runtimes = {
+        item.nodeid: runtimes[item.nodeid] if item.nodeid in runtimes else sys.maxsize
+        for item in pytest_items
+    }
```
Comment (alexey-vyskubov) [inline pytest_rts/utils/common.py:110]: here you get second iteration over pytest_items; why not
```
return sorted(selected, key = lambda item: item.nodeid if item.nodeid in runtimes else sys.maxsize)
```

File: tox.ini [hunk 3]
```diff
@@ -10,7 +10,7 @@ commands =
         pylint -j 0 -s n setup.py pytest_rts
         mypy setup.py pytest_rts --ignore-missing-imports
         safety check
-        pytest \
+        python3 -m pytest \
```
Comment (alexey-vyskubov) [inline tox.ini:13]: what's the purpose of this change?

**Review by alexey-vyskubov — COMMENTED** [review]
Looks good in general; few minor comments.

File: pytest_rts/pytest/mapper_plugin.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                found_testfiles=self.testfiles,
+                test_function_lines=self.test_func_lines,
+            )
+            self.mappinghelper.save_testrun_data(testrun_data)
+        else:
+            yield
+
+    def pytest_sessionfinish(self, session):
+        """End SQLalchemy transaction after test loop"""
+        del session
```
[the remark this answers is not in the document]
Comment (guotin) [reply pytest_rts/pytest/mapper_plugin.py:53]: This was my answer to get rid of pylint error for unused argument. I copied all the parameters of functions from the pytest api page even if they're not used. Is there a better way to handle something like this?

[the remark this answers is not in the document]
Comment (alexey-vyskubov) [reply pytest_rts/pytest/mapper_plugin.py:53]: Just name the parameter with underscore in the beginning, pylint will be happy, and people reading the code will know it's meant to be unused: `def pytest_sessionfinish(self, _session)`.

[the remark this answers is not in the document]
Comment (alexey-vyskubov) [reply pytest_rts/pytest/mapper_plugin.py:53]: What you do is dangerous, see:
```
In [1]: class Foo:
   ...:     def __del__(self):
   ...:         print("Unexpected side effect")
   ...: 

In [2]: def f(foo):
   ...:     del foo
   ...:     # this function does nothing, right?

In [3]: f(Foo())
Unexpected side effect
```

[the remark this answers is not in the document]
Comment (guotin) [reply pytest_rts/pytest/mapper_plugin.py:53]: Pytest doesn't seem to like the underscores. For example if I change `pytest_collection_modifyitems(self, session, config, items)` to `pytest_collection_modifyitems(self, _session, _config, items)` it gives me this:

```
INTERNALERROR> pluggy.manager.PluginValidationError: Plugin 'rts-init-plugin' for hook 'pytest_collection_modifyitems'
INTERNALERROR> hookimpl definition: pytest_collection_modifyitems(_session, _config, items)
INTERNALERROR> Argument(s) {'_config', '_session'} are declared in the hookimpl but can not be found in the hookspec
```

Removing the unused `config` and `session` completely seems to work but does that cause some problems?

File: tox.ini (hunk 3 above)
Comment (guotin) [reply tox.ini:13]: After adding `__init__.py` to the `pytest_rts/tests/` folder I can't get tox working without changing `pytest` to `python3 -m pytest`. Causes some import error and I couldn't find any other solution

File: pytest_rts/pytest/mapper_plugin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexey-vyskubov) [reply pytest_rts/pytest/mapper_plugin.py:53]: Okay, in this case just add pylint: disable; don't del what you receive.

File: tox.ini (hunk 3 above)
Comment (alexey-vyskubov) [reply tox.ini:13]: It would be nice to understand the root cause before making random changes :)

Comment (guotin) [reply tox.ini:13]: Import fails for finding `pytest_rts.tests` module in `tox` so would that module being missing from `PYTHONPATH` be the root cause? Another fix I found would be to the following to `tox.ini`:
```
setenv =
    PYTHONPATH = {toxinidir}
```

This also seems to collect coverage and adding `* /.tox/*` to `.coveragerc` to omit the duplicate files results in the following coverage report inside tox

``` 
----------- coverage: platform linux, python 3.8.5-final-0 -----------
Name                                       Stmts   Miss  Cover
--------------------------------------------------------------
pytest_rts/collect.py                          8      8     0%
pytest_rts/connection.py                      12      0   100%
pytest_rts/plugin.py                          68      7    90%
pytest_rts/pytest/collect_plugin.py           13     13     0%
pytest_rts/pytest/fake_item.py                 3      0   100%
pytest_rts/pytest/init_phase_plugin.py        12      0   100%
pytest_rts/pytest/mapper_plugin.py            29      3    90%
pytest_rts/pytest/normal_phase_plugin.py      11      6    45%
pytest_rts/pytest/update_phase_plugin.py      22      2    91%
pytest_rts/utils/common.py                    61      4    93%
pytest_rts/utils/git.py                       64      4    94%
pytest_rts/utils/mappinghelper.py            103      1    99%
pytest_rts/utils/selection.py                 38      0   100%
pytest_rts/utils/tables.py                     8      0   100%
pytest_rts/utils/testgetter.py                61      5    92%
--------------------------------------------------------------
TOTAL                                        513     53    90%
Coverage HTML written to dir htmlcov


=============================================================================================== 18 passed in 38.63s ===============================================================================================
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
  py38: commands succeeded
  congratulations :)
``` 

Does this look okay?
